### PR TITLE
Download and convert model

### PR DESCRIPTION
**Pull request by likholat**

### CODE REVIEW

**Comment by likholat** [comment]
@dkurt

File: tools/downloader/topologies.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import sys
+import common
+import downloader
+import converter
+
+_cache = Path(os.getenv('OPEN_MODEL_ZOO_CACHE_DIR', Path('.').resolve() / 'models'))
+
+def load(model_name):
+    downloader.main(['--name=' + model_name, '--output_dir=' + str(_cache).replace('\\', '\\\\')])
```
Comment (dkurt) [inline tools/downloader/topologies.py:13]: Let's create `Zoo` class which can provide some context and have supportive methods (i.e. list all the models names). So method can be named as `load_model` (see https://github.com/openvinotoolkit/open_model_zoo/pull/304#issuecomment-567436663)

Comment (dkurt) [reply tools/downloader/topologies.py:13]: please ignore it

File: tools/downloader/topologies.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+from pathlib import Path
+
+import sys
+import common
+import downloader
+import converter
+
+_cache = Path(os.getenv('OPEN_MODEL_ZOO_CACHE_DIR', Path('.').resolve() / 'models'))
+
+def load(model_name):
```
Comment (dkurt) [inline tools/downloader/topologies.py]: `load_model`

File: tools/downloader/topologies.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+from pathlib import Path
+
+import common
+import downloader
+import converter
+
+_cache = Path(os.getenv('OPEN_MODEL_ZOO_CACHE_DIR', Path('.').resolve() / 'models'))
+
+class Topology:
+    def load_model(self, model_name, precision=None):
```
Comment (dkurt) [inline tools/downloader/topologies.py]: Why this method is inside Topology class? Should be standalone

**Comment by dkurt** [comment]
Please keep in the description some code snippet with an example how to use proposed API

File: tools/downloader/topologies_test.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import unittest
+import os
+
+from conftest import download_dir_path, image_path
+from topologies import Topology, load_model
+
+class TestTopologies(unittest.TestCase):
+    download_path = download_dir_path()
+    ssd = load_model('mobilenet-ssd', 'FP32')
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: FP32 should be by default

Comment (dkurt) [inline tools/downloader/topologies_test.py]: Please make import:
```
import topologies as omz

ssd = omz.load_model('mobilenet-ssd', 'FP32')
```

File: tools/downloader/topologies_test.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    def test_get_ir(self):
+        xmlPath, binPath = self.ssd.get_ir()
+        self.assertEqual(xmlPath, 
+            os.path.join(self.download_path, 'public', 'mobilenet-ssd', 'FP32', 'mobilenet-ssd.xml'))
+        self.assertEqual(binPath, 
+            os.path.join(self.download_path, 'public', 'mobilenet-ssd', 'FP32', 'mobilenet-ssd.bin'))
+
+    def test_get_ie(self):
+        from openvino.inference_engine import IENetwork
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: Why do we need this import?

File: tools/downloader/topologies_test.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        self.assertTrue(isinstance(net, IENetwork))
+
+    def test_result_model(self):
+        import cv2 as cv
+        import numpy as np
+        from openvino.inference_engine import IECore
+
+        ie = IECore()
+        net = self.ssd.get_ie_network()
+        exec_net = ie.load_network(network=net, device_name='CPU')
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: Please avoid extra parameters (`load_network(network=net)`). Just
```
exec_net = ie.load_network(net, 'CPU')
```

File: tools/downloader/topologies.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+def load_model(model_name, precision=None):
+        parser = argparse.ArgumentParser()
+        parser.add_argument('--name', metavar='PAT[,PAT...]',
+            help='only dump info for models whose names match at least one of the specified patterns')
+        parser.add_argument('--list', type=Path, metavar='FILE.LST',
+            help='only dump info for models whose names match at least one of the patterns in the specified file')
+        parser.add_argument('--all', action='store_true', help='dump info for all available models')
+        parser.add_argument('--print_all', action='store_true', help='print all available models')
+        args = parser.parse_args(['--name=' + model_name])
```
Comment (dkurt) [inline tools/downloader/topologies.py]: Why we need Argument parser?

File: tools/downloader/topologies_test.py (hunk 5 above)
Comment (likholat) [reply tools/downloader/topologies_test.py]: We need this import for test: ```self.assertTrue(isinstance(net, IENetwork))```
Else ```NameError: name 'IENetwork' is not defined```

File: tools/downloader/topologies.py (hunk 7 above)
Comment (likholat) [reply tools/downloader/topologies.py]: We use it in ```common.load_models_from_args(parser, args)```

File: tools/downloader/topologies.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+_cache = Path(os.getenv('OPEN_MODEL_ZOO_CACHE_DIR', Path('.').resolve() / 'models'))
+
+def __load(model_name):
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--name', metavar='PAT[,PAT...]',
+        help='only dump info for models whose names match at least one of the specified patterns')
+    parser.add_argument('--list', type=Path, metavar='FILE.LST',
+        help='only dump info for models whose names match at least one of the patterns in the specified file')
+    parser.add_argument('--all', action='store_true', help='dump info for all available models')
+    parser.add_argument('--print_all', action='store_true', help='print all available models')
```
Comment (dkurt) [inline tools/downloader/topologies.py]: please remove unused code

File: tools/downloader/topologies.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    args = Namespace(all=False, list=None, name='mobilenet-ssd', print_all=False)
```
Comment (dkurt) [inline tools/downloader/topologies.py]: hardcoded name

File: tools/downloader/topologies.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                '--output_dir=' + str(_cache).replace('\\', '\\\\')])
+        
+    return Topology(model_name, download_dir, precision)
+
+def help(model_name):
+    _topology = __load(model_name)
+    description_res = 'Model description for ' + model_name +' \n  ' + \
+                _topology.description + '\n     License: ' +  _topology.license_url
+
+    print(description_res)
```
Comment (dkurt) [inline tools/downloader/topologies.py]: Please show example of `help(ssd)`

File: tools/downloader/topologies.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+class Topology:
+    def __init__(self, name, download_dir, precision):
+        self.name = name
+        self.download_dir = download_dir
+        self.precision = precision
+   
+    def get_ir(self, precision=None):
```
Comment (dkurt) [inline tools/downloader/topologies.py]: FP32 by default

File: tools/downloader/topologies.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        precision = precision if precision else self.precision
+        prefix = Path(self.download_dir) / precision / self.name
+
+        xmlPath = str(prefix) + '.xml'
+        binPath = str(prefix) + '.bin'
+        if os.path.exists(xmlPath) and os.path.exists(binPath):
+            return xmlPath, binPath
+
+        converter.main(['--name=' + self.name, '--precisions=' + precision,
+                        '--download_dir=' + str(_cache).replace('\\', '\\\\')])
```
Comment (dkurt) [inline tools/downloader/topologies.py:79]: Is there `download_dir` for conversion? I thought there is only `output_dir`

File: tools/downloader/topologies_test.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    download_path = os.environ.get("OPEN_MODEL_ZOO_CACHE_DIR")
+    ssd = omz.load_model('mobilenet-ssd')
+
+    def test_load_model(self):
+        self.assertTrue(os.path.exists(
+            os.path.join(self.download_path, 'public', 'mobilenet-ssd', 'mobilenet-ssd.caffemodel')))
+
+    def test_get_ir(self):
+        xmlPath, binPath = self.ssd.get_ir()
+        self.assertTrue(os.path.exists(self.download_path))
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: Why we test download_dir? We need to test that model files exist:
```python
self.assertTrue(os.path.exists(xmkPath))
self.assertTrue(os.path.exists(binPath))
```

File: tools/downloader/topologies_test.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+    download_path = os.environ.get("OPEN_MODEL_ZOO_CACHE_DIR")
+    ssd = omz.load_model('mobilenet-ssd')
+
+    def test_load_model(self):
+        self.assertTrue(os.path.exists(
+            os.path.join(self.download_path, 'public', 'mobilenet-ssd', 'mobilenet-ssd.caffemodel')))
+
+    def test_get_ir(self):
+        xmlPath, binPath = self.ssd.get_ir()
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: `xml_path` and `bin_path` (snake_case is preferable for Python)

File: tools/downloader/topologies.py (hunk 10 above)
Comment (likholat) [reply tools/downloader/topologies.py]: image

File: tools/downloader/topologies_test.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        exec_net = ie.load_network(net, 'CPU')
+
+        image = cv.imread(image_path)
+        rows = image.shape[0]
+        cols = image.shape[1]
+        
+        image_size_x = 300
+        image_size_y = 300 
+
+        img = cv.resize(image, (image_size_x, image_size_y))
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: just
```
img = cv.resize(image, (300, 300))
```

File: tools/downloader/topologies.py (hunk 10 above)
Comment (likholat) [reply tools/downloader/topologies.py]: We  should use it as ```omz.help('model-name')```

File: tools/downloader/topologies_test.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+            xmin = int(predictions[0][0][curProposal][3] * cols)
+            ymin = int(predictions[0][0][curProposal][4] * rows)
+            xmax = int(predictions[0][0][curProposal][5] * cols)
+            ymax = int(predictions[0][0][curProposal][6] * rows)
+
+        self.assertEqual(label, 7)
+        self.assertEqual(xmin, 266)
+        self.assertEqual(ymin, 88)
+        self.assertEqual(xmax, 449)
+        self.assertEqual(ymax, 379)
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: Better to test IoU in relative coordinates

File: tools/downloader/topologies.py (hunk 10 above)
Comment (dkurt) [reply tools/downloader/topologies.py]: But what about help for `class Topology`?

File: tools/downloader/topologies.py (hunk 12 above)
Comment (likholat) [reply tools/downloader/topologies.py:79]: image

Comment (dkurt) [reply tools/downloader/topologies.py:79]: I see, thanks!

File: tools/downloader/topologies.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+def load_model(model_name, precision='FP32'):
+    download_dir = _cache / __load(model_name).subdirectory
+
+    prototxt_num = 0
+    for path in Path(download_dir).rglob('*.prototxt'):
+        prototxt_num += 1
+
+    caffemodel_num = 0
+    for path in Path(download_dir).rglob('*.caffemodel'):
+        caffemodel_num += 1
```
Comment (dkurt) [inline tools/downloader/topologies.py]: Please try to use downloader's cache instead

File: tools/downloader/topologies_test.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        img = np.expand_dims(img, axis=0)
+        img = img.transpose((0, 3, 1, 2))
+
+        input_blob = next(iter(net.input_info))
+        out_blob = next(iter(net.outputs))
+
+        predictions = exec_net.infer(inputs={input_blob: img})
+        predictions = predictions[out_blob]
+
+        for curProposal in range(predictions.shape[2]):
```
Comment (dkurt) [inline tools/downloader/topologies_test.py]: Pleae use for each:
```python
for detection in predictions.reshape(-1, 7):
```

File: tools/downloader/topologies.py (hunk 10 above)
Comment (likholat) [reply tools/downloader/topologies.py]: image

**Comment by dkurt** [comment]
see https://github.com/openvinotoolkit/open_model_zoo/pull/1352

**Comment by dkurt** [comment]
please close this PR

**Pull request closed without merging**